In [1]:
import numpy as np
import pandas as pd
from sklearn.metrics import accuracy_score  
from sklearn.metrics import recall_score  
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.neural_network import MLPClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.linear_model import LogisticRegression 
from sklearn import neighbors
from sklearn import tree
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import BernoulliNB
import matplotlib.pyplot as plt

In [2]:
def get_the_predicted_bugs(name, num1, num2):
    All_Data = pd.read_csv(name, usecols = [i for i in range(num1, num2)]).values
    return All_Data

In [3]:
data = get_the_predicted_bugs('modified_all_data.csv', 2, 45)

## Logistic Regression

## best accuracy_score

In [4]:
data_random = np.random.permutation(data)
x = data_random[:, [i for i in range(data.shape[1] - 1)]]
y = data_random[:, -1]
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.4, random_state = 0)
lr_clf = LogisticRegression()

In [5]:
param_grid = [
             {
                 'penalty' : ['l1'],
                 'C' : [0.001, 0.01, 0.1, 1, 10, 100],
                 'solver' : ['liblinear']
             },
             {
                 'penalty' : ['l2'],
                 'C' : [0.001, 0.01, 0.1, 1, 10, 100],
                 'solver' : ['newton-cg', 'lbfgs', 'sag', 'liblinear']
             }
             ]
grid_search = GridSearchCV(lr_clf, param_grid, scoring = 'accuracy', n_jobs = -1, cv = 3)

In [6]:
grid_search.fit(x_train, y_train)

GridSearchCV(cv=3, error_score='raise',
       estimator=LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False),
       fit_params=None, iid=True, n_jobs=-1,
       param_grid=[{'penalty': ['l1'], 'C': [0.001, 0.01, 0.1, 1, 10, 100], 'solver': ['liblinear']}, {'penalty': ['l2'], 'C': [0.001, 0.01, 0.1, 1, 10, 100], 'solver': ['newton-cg', 'lbfgs', 'sag', 'liblinear']}],
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring='accuracy', verbose=0)

In [7]:
grid_search.best_params_

{'C': 0.01, 'penalty': 'l1', 'solver': 'liblinear'}

In [8]:
grid_search.best_estimator_

LogisticRegression(C=0.01, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l1', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [9]:
y_pre = grid_search.best_estimator_.predict(x_test)
accuracy_score(y_test, y_pre), recall_score(y_test, y_pre)

(0.8781302170283807, 0.19230769230769232)

## get the average accuracy_score

In [24]:
accuracy_score_list = []
recall_score_list = []
for i in range(1000):
    data_random = np.random.permutation(data)
    x = data_random[:, [i for i in range(data.shape[1] - 1)]]
    y = data_random[:, -1]
    x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.4, random_state = 0)
    lr_clf = LogisticRegression(C = 0.01, penalty = 'l1', solver = 'liblinear')
    lr_clf.fit(x_train, y_train)
    y_pre = lr_clf.predict(x_test)
    accuracy_score_list.append(accuracy_score(y_test, y_pre))
    recall_score_list.append(recall_score(y_test, y_pre))

In [25]:
np.mean(accuracy_score_list)

0.8653539232053422

In [26]:
np.mean(recall_score_list)

0.14058533043248342

By Logistic Regression, aftering tuning the parameters for getting the best accuracy_score, the average accuracy_score of 1000 times is about is about 0.865 and the average recall_score of 1000 times is about 0.14.

## best recall_score

In [37]:
data_random = np.random.permutation(data)
x = data_random[:, [i for i in range(data.shape[1] - 1)]]
y = data_random[:, -1]
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.4, random_state = 0)
lr_clf = LogisticRegression()

In [38]:
param_grid = [
             {
                 'penalty' : ['l1'],
                 'C' : [0.001, 0.01, 0.1, 1, 10, 100],
                 'solver' : ['liblinear']
             },
             {
                 'penalty' : ['l2'],
                 'C' : [0.001, 0.01, 0.1, 1, 10, 100],
                 'solver' : ['newton-cg', 'lbfgs', 'sag', 'liblinear']
             }
             ]
grid_search = GridSearchCV(lr_clf, param_grid, scoring = 'recall', n_jobs = -1, cv = 3)

In [39]:
grid_search.fit(x_train, y_train)

GridSearchCV(cv=3, error_score='raise',
       estimator=LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False),
       fit_params=None, iid=True, n_jobs=-1,
       param_grid=[{'penalty': ['l1'], 'C': [0.001, 0.01, 0.1, 1, 10, 100], 'solver': ['liblinear']}, {'penalty': ['l2'], 'C': [0.001, 0.01, 0.1, 1, 10, 100], 'solver': ['newton-cg', 'lbfgs', 'sag', 'liblinear']}],
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring='recall', verbose=0)

In [40]:
grid_search.best_params_

{'C': 10, 'penalty': 'l2', 'solver': 'lbfgs'}

In [41]:
y_pre = grid_search.best_estimator_.predict(x_test)
accuracy_score(y_test, y_pre), recall_score(y_test, y_pre)

(0.8464106844741235, 0.225)

## get the average accuracy_score

In [43]:
accuracy_score_list = []
recall_score_list = []
for i in range(1000):
    data_random = np.random.permutation(data)
    x = data_random[:, [i for i in range(data.shape[1] - 1)]]
    y = data_random[:, -1]
    x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.4, random_state = 0)
    lr_clf = LogisticRegression(C = 10, penalty = 'l2', solver = 'lbfgs')
    lr_clf.fit(x_train, y_train)
    y_pre = lr_clf.predict(x_test)
    accuracy_score_list.append(accuracy_score(y_test, y_pre))
    recall_score_list.append(recall_score(y_test, y_pre))

In [44]:
np.mean(accuracy_score_list)

0.8636961602671117

In [45]:
np.mean(recall_score_list)

0.1758725280844579

By Logistic Regression, aftering tuning the parameters for getting the best recall_score, the average accuracy_score of 1000 times is about is about 0.86 and the average recall_score of 1000 times is about 0.176.